__tests__/get-options-overrides.spec.ts

```diff
@@ -0,0 +1,164 @@
+	tsconfigOverride: null,
+	useTsconfigDeclarationDir: false,
+	tsconfigDefaults: null,
+	sourceMapCallback: (id: string, map: string): void => {
+		console.log(id + map);
+	},
+	transformers: [(ls: ts.LanguageService) => {
+		console.log(ls);
+		return {};
+	}],
+};
+
+const forcedOptions: ts.CompilerOptions = {
+	allowNonTsExtensions: true,
+	importHelpers: true,
+	inlineSourceMap: false,
+	moduleResolution: ts.ModuleResolutionKind.NodeJs,
+	noEmit: false,
+	noEmitHelpers: false,
+	noResolve: false,
+	outDir: "placeholder", // normalized
+}
+
+const defaultPreParsedTsConfig: ts.ParsedCommandLine = {
+	options: {},
+	fileNames: [],
+	errors: [],
+};
+
+test("getOptionsOverrides", () => {
+	const config = { ...defaultConfig };
+	expect(normalizePaths(["outDir"], getOptionsOverrides(config))).toStrictEqual(
+		{
+			...forcedOptions,
+		},
+	);
+});
+
+test("getOptionsOverrides - preParsedTsConfig", () => {
+	const config = { ...defaultConfig };
+	const preParsedTsConfig = { ...defaultPreParsedTsConfig };
+	expect(normalizePaths(["outDir"], getOptionsOverrides(config, preParsedTsConfig))).toStrictEqual(
+		{
+			...forcedOptions,
+			declarationDir: undefined,
+			module: ts.ModuleKind.ES2015,
+			sourceRoot: undefined,
+		},
+	);
+});
+
+test("getOptionsOverrides - preParsedTsConfig with options.module", () => {
+	const config = { ...defaultConfig };
+	const preParsedTsConfig = {
+		...defaultPreParsedTsConfig,
+		options: {
+			module: ts.ModuleKind.AMD,
+		},
+	};
+	expect(normalizePaths(["outDir"], getOptionsOverrides(config, preParsedTsConfig))).toStrictEqual(
+		{
+			...forcedOptions,
+			declarationDir: undefined,
+			sourceRoot: undefined,
+		},
+	);
+});
+
+test("getOptionsOverrides - with declaration", () => {
+	const config = { ...defaultConfig, useTsconfigDeclarationDir: true };
+	const preParsedTsConfig = { ...defaultPreParsedTsConfig };
+	expect(normalizePaths(["outDir"], getOptionsOverrides(config, preParsedTsConfig))).toStrictEqual(
+		{
+			...forcedOptions,
+			module: ts.ModuleKind.ES2015,
+			sourceRoot: undefined,
+		},
+	);
+});
+
+test("getOptionsOverrides - with sourceMap", () => {
+	const config = { ...defaultConfig };
+	const preParsedTsConfig = {
+		...defaultPreParsedTsConfig,
+		options: {
+			sourceMap: true,
+		},
+	};
+	expect(normalizePaths(["outDir"], getOptionsOverrides(config, preParsedTsConfig))).toStrictEqual(
+		{
+			...forcedOptions,
+			declarationDir: undefined,
+			module: ts.ModuleKind.ES2015,
+		},
+	);
+});
+
+test("createFilter", () => {
+	const config = { ...defaultConfig };
+	const preParsedTsConfig = { ...defaultPreParsedTsConfig };
+
+	const stubbedContext = makeStubbedContext({});
+	const filter = createFilter(stubbedContext, config, preParsedTsConfig);
+
+	expect(filter("src/test.ts")).toBe(true);
+	expect(filter("src/test.js")).toBe(false);
+	expect(filter("src/test.d.ts")).toBe(false);
+});
+
+// not totally sure why this is failing
+test.skip("createFilter -- rootDirs", () => {
```

Comment:
So as far as I understand, I believe I'm testing the `expandIncludeWithDirs` function correctly.

And according to the original PR that wrote the code, #38, this looks like the intended usage too.
(I wrote the test before looking at that PR, fwiw)

Reply:
@ezolenko do you have any idea why this test fails from a quick glance? would love to get it working; don't think I can blame TS internals for this one 😆

Reply:
Looks like this has nothing to do with `rootDirs`, rollup filter just doesn't like folder prefixes. This test fails as well:

```test("createFilter -- rootDirs 2", () => {
	const config = {
		...defaultConfig,
		include : ["src/*.ts+(|x)"],
	};
	const preParsedTsConfig = {
		...defaultPreParsedTsConfig,
	};

	const stubbedContext = makeStubbedContext({});
	const filter = createFilter(stubbedContext, config, preParsedTsConfig);

	expect(filter("src/test.ts")).toBe(true);
});```

Reply:
`include : ["src/test.ts"],` also fails, but `include : ["**/src/test.ts"],` succeeds. `include : ["/src/test.ts"] [...] filter("/src/test.ts")` also works.

This likely has something to do with the way `rootDirs` are presented in typescript after parsing tsconfig -- I can't see this value in `parsedConfig.options.rootDirs` at runtime (even though it is there in types and in tsconfig), but I see it in `parsedConfig.wildcardDirectories` as an absolute path.

I think the bug is that we are using `parsedConfig.options.rootDirs` instead of `parsedConfig.wildcardDirectories`. If we switch, rollup will get absolute paths in includes and things would work.

Reply:
I wonder if typescript deprecated `CompilerOptions.rootDirs` value or stopped writing it sometime after this functionality was implemented.

Reply:
So [`rootDirs`](https://www.typescriptlang.org/tsconfig#rootDirs) is definitely still used by TS; I remember in https://github.com/jaredpalmer/tsdx/issues/668 people were using `rootDirs` as a workaround with Storybook, which is only 2 years old.

I did see that `@rollup/pluginutils` has had some [breaking changes in v4.0.0](https://github.com/rollup/plugins/blob/master/packages/pluginutils/CHANGELOG.md#v400) to `createFilter`, but I don't think that impacts our usage here... (It is perhaps different than what the author of #38 originally intended though?)

It seems like the problem with these tests is that `filter` actually generally takes an _absolute_ path. The usage of `filter` in the codebase seems to match that too -- only absolute paths are used with it.
The [docs](https://github.com/rollup/plugins/tree/master/packages/pluginutils#createfilter) for `pluginutils` say that `createFilter` resolves the `include`/`exclude` paths with `process.cwd()` (when `**` is not given as of v4, since that means all dirs), and I do see that in [the code](https://github.com/rollup/plugins/blob/556837f0b7e7323f237370e36394569abc2b1ef9/packages/pluginutils/src/createFilter.ts#L15).

So I actually think what's needed to get these tests to pass is just to concat the path with `process.cwd()` so it has an absolute path equivalent

Reply:
Got it working in #329